async_sendgrid/sendgrid.py

```diff
@@ -1,84 +1,115 @@
-            by "On-Behalf-Of" header by underlying client.
-            See https://sendgrid.com/docs/User_Guide/Settings/subusers.html
-            for more details
-        :type impersonate_subuser: string
-        :param host: base URL for API calls
-        :type host: string
-        """
+        self._api_key = api_key
+        self._endpoint = "https://api.sendgrid.com/v3/mail/send"
+        self._version = async_sendgrid.__version__
 
-        self.host = host
-        self.session = session
-        self.timeout = timeout
-        self.version = async_sendgrid.__version__
-        self.useragent = 'async_sendgrid/{};python'.format(self.version)
-        self.headers = {
-            "Authorization": f"Bearer {api_key}",
-            "User-Agent": self.useragent,
+        self._headers = {
+            "Authorization": f"Bearer {self._api_key}",
+            "User-Agent": f"async_sendgrid/{self._version};python",
             "Accept": "*/*",
             "Content-Type": "application/json",
         }
+
         if impersonate_subuser:
-            self.headers['On-Behalf-Of'] = impersonate_subuser
+            self._headers["On-Behalf-Of"] = impersonate_subuser
 
-    async def __aenter__(self):
-        await self.open()
-        return self
+        self._session = create_session(self._headers)
+
+    @property
+    def api_key(self) -> str:
+        return self._api_key
 
-    async def __aexit__(self, exc_type, exc, tb):
-        await self.close()
+    @property
+    def endpoint(self) -> str:
+        return self._endpoint
 
-    async def open(self):
-        if self.session is None:
-            self.session = httpx.AsyncClient(
-                headers=self.headers,
-            )
+    @property
+    def version(self) -> str:
+        return self._version
 
-    async def close(self):
-        if self.session is not None:
-            await self.session.aclose()
-            self.session = None
+    @property
+    def headers(self) -> dict[Any, Any]:
+        return self._headers
 
-    async def send(self, message: dict):
+    async def send(self, message: Mail) -> Response:
         """
         Make a Twilio SendGrid v3 API request with the request body generated
         by the Mail object
 
-        :param message: The Twilio SendGrid v3 API request body generated
-            by the Mail object or dict
+        Parameters:
+        ----
+            :param message: The Twilio SendGrid v3 API request body generated
+                by the Mail object or dict
+        Returns:
+        ----
+            :return: The Twilio SendGrid v3 API response
         """
-        url = f"{self.host}/mail/send"
-        if not isinstance(message, dict):
-            message = message.get()
+        json_message = message.get()
+        response = await self._session.post(
+            url=self._endpoint, json=json_message
+        )
+        return response
```

Comment:
In my opinion, you should add support not only for sendgrid.Mail, but also for dict[Any, Any], because there will be users who will not want to install the sendgrid package in their project.


```python
    async def send(self, message: Mail | dict[str, Any]) -> Response:
        """
        Make a Twilio SendGrid v3 API request with the request body generated
        by the Mail object

        Parameters:
        ----
            :param message: The Twilio SendGrid v3 API request body generated
                by the Mail object or dict

        Returns:
        ----
            :return: The Twilio SendGrid v3 API response
        """
        if not isinstance(message, dict):
            json_message = message.get()
        response = await self._session.post(
            url=self._endpoint, json=json_message
        )
        return response
```

Reply:
It is 100% understandable. Meanwhile, I am afraid that it will break sometimes if nobody looks after the specific form they request (which we don't have to if we use their maintained Mail's api).

Reply:
I thought about it and decided you're right, so I'm closing this conversation, but I think that in the future we should implement our own class, analogue to Mail, to implement this idea